Firestore/core/test/unit/util/md5_test.cc

```diff
@@ -0,0 +1,170 @@
+
+// Returns the md5 digest for the given string.
+// This function does not "calculate" the digest, but rather has a hardcoded set
+// of pre-calculated digests that it returns.
+// It is an error if this function does not have a pre-calculated digest for the
+// given string.
+std::array<unsigned char, 16> GetPreComputedMd5Digest(const std::string&);
+
+// Generates and returns a string with all possible characters with the given
+// length. The given length must be at least 256.
+std::string GetStringWithAllPossibleCharacters(int length);
+
+TEST(CalculateMd5DigestTest, ShouldReturnMd5DigestOfEmptyString) {
+  EXPECT_EQ(CalculateMd5Digest(""), GetPreComputedMd5Digest(""));
+}
+
+TEST(CalculateMd5DigestTest, ShouldReturnMd5DigestOfA) {
+  EXPECT_EQ(CalculateMd5Digest("a"), GetPreComputedMd5Digest("a"));
+}
+
+TEST(CalculateMd5DigestTest, ShouldReturnMd5DigestOfABC) {
+  EXPECT_EQ(CalculateMd5Digest("abc"), GetPreComputedMd5Digest("abc"));
+}
+
+TEST(CalculateMd5DigestTest, ShouldReturnMd5DigestOfHelloWorld) {
+  EXPECT_EQ(CalculateMd5Digest("hello world!"),
+            GetPreComputedMd5Digest("hello world!"));
+}
+
+TEST(CalculateMd5DigestTest, ShouldReturnMd5DigestOfTheQuickBrownFox) {
+  EXPECT_EQ(
+      CalculateMd5Digest("the quick brown fox jumps over the lazy dog"),
+      GetPreComputedMd5Digest("the quick brown fox jumps over the lazy dog"));
+}
+
+TEST(CalculateMd5DigestTest, ShouldReturnMd5DigestOfShortStringWithAllChars) {
+  const std::string s = GetStringWithAllPossibleCharacters(512);
+  EXPECT_EQ(CalculateMd5Digest(s), GetPreComputedMd5Digest(s));
+}
+
+TEST(CalculateMd5DigestTest, ShouldReturnMd5DigestOfLongStringWithAllChars) {
+  const std::string s = GetStringWithAllPossibleCharacters(8192);
+  EXPECT_EQ(CalculateMd5Digest(s), GetPreComputedMd5Digest(s));
+}
+
+unsigned char UnsignedCharFromHexDigit(char digit) {
+  switch (digit) {
+    case '0':
+      return 0x0;
+    case '1':
+      return 0x1;
+    case '2':
+      return 0x2;
+    case '3':
+      return 0x3;
+    case '4':
+      return 0x4;
+    case '5':
+      return 0x5;
+    case '6':
+      return 0x6;
+    case '7':
+      return 0x7;
+    case '8':
+      return 0x8;
+    case '9':
+      return 0x9;
+    case 'a':
+      return 0xA;
+    case 'b':
+      return 0xB;
+    case 'c':
+      return 0xC;
+    case 'd':
+      return 0xD;
+    case 'e':
+      return 0xE;
+    case 'f':
+      return 0xF;
+  }
+  HARD_FAIL("unrecognized hex digit: %s", std::to_string(digit));
+}
+
+std::array<unsigned char, 16> UnsignedCharArrayFromHexDigest(
+    const std::string& s) {
+  HARD_ASSERT(s.length() == 32);
+  std::array<unsigned char, 16> result;
+  for (int i = 0; i < 16; ++i) {
+    unsigned char c1 = UnsignedCharFromHexDigit(s[i * 2]);
+    unsigned char c2 = UnsignedCharFromHexDigit(s[(i * 2) + 1]);
+    result[i] = (c1 << 4) | c2;
+  }
+  return result;
+}
+
+std::array<unsigned char, 16> GetPreComputedMd5Digest(const std::string& s) {
```

Comment:
Can we add special characters test case as well: https://github.com/firebase/firebase-js-sdk/blob/97ca3ee1580ca8b90c3293da2652eccf02709030/packages/firestore/test/unit/core/webchannel_wrapper.test.ts#L62

Reply:
But GetStringWithAllPossibleCharacters should have covered special characters here. 🤔

Reply:
Great idea. But in C++ strings are just "byte strings" that know nothing about Unicode code points/code units and surrogate characters and stuff like that. Since MD5 only processes bytes (not characters) there probably isn't much value in adding tests for UTF-8 encoding of Unicode characters.